packages/optimizer/lib/transformers/PreloadHeroImage.js

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {createElement, nextNode, insertAfter, firstChildByTag} = require('../NodeUtils');
+const {findMetaViewport} = require('../HtmlDomHelper');
+const parseSrcSet = require('../parseSrcSet');
+
+// Images smaller than 150px are considered tiny
+const TINY_IMG_THRESHOLD = 150;
+
+/**
+ * PreloadHeroImage - This transformer identifies a hero image or
+ * important images on the document and attaches <link rel="preload"> element.
+ *
+ * This transformer supports the following option:
+ *
+ * * `preloadHeroImage`: [true|false] - enables or disables hero image preloading. The default is `true`.
+ */
+class PreloadHeroImage {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = config.preloadHeroImage !== false && config.experimentPreloadHeroImage;
+    this.log.info('enabled', this.enabled);
+  }
+  async transform(root, params) {
+    if (!this.enabled || params.preloadHeroImage === false) {
+      return;
+    }
+    const html = firstChildByTag(root, 'html');
+    const head = firstChildByTag(html, 'head');
+    const body = firstChildByTag(html, 'body');
+    if (!body || !head) return;
+
+    const heroImage = this.findHeroImage(body);
+
+    if (!heroImage) {
+      return;
+    }
+    let referenceNode = findMetaViewport(head);
+
+    const preload = createElement('link', {
+      'rel': 'preload',
+      'href': heroImage.src,
+      'as': 'image',
+      'data-hero': '',
+    });
+    if (heroImage.srcset) {
+      try {
+        parseSrcSet(heroImage.srcset);
+        preload.attribs.srcset = heroImage.srcset;
+      } catch (err) {
+        this.log.warn(
+          `Invalid srcset '${heroImage.srcset}' for amp-img src '${heroImage.src}'`,
+          err.message
+        );
+      }
+    }
+    this.log.debug('Preloading hero image: ', heroImage.src);
+    insertAfter(head, preload, referenceNode);
+  }
+
+  findHeroImage(root) {
+    if (!root.tagName) {
+      return null;
+    }
+    // Ignore images inside templates
+    if (root.tagName === 'template') {
+      return null;
+    }
+
+    const layout = root.attribs ? root.attribs.layout : '';
```

Comment:
purely my own edification, but what are the supported engines for this project? I was thinking Optional Chaining (layout = root.attribs?.layout || '') could be used here, but it is only in node 14+. Obviously not ok for the project, but I am not sure what the projects attempts to suport

Reply:
nevermind, remembered .travis.yml existed. node >=v10